gtk/gtkfilechooserdialog.c

```diff
@@ -507,6 +509,61 @@ add_button (GtkWidget *button, gpointer data)
 }
 
 static void
+setup_places (GtkFileChooserDialog *dialog)
+{
+  gboolean use_header;
+
+  if (dialog->priv->places_setup)
+    return;
+
+  dialog->priv->places_setup = TRUE;
+
+  g_object_get (dialog, "use-header-bar", &use_header, NULL);
+  if (use_header)
+    {
+      GtkWidget *button;
+      GtkWidget *image;
+      GtkWidget *header;
+      GdkScreen *screen;
+      GtkPlacesSidebar *sidebar;
+      GtkRequisition sidebar_requisition;
+      gboolean sidebar_visible;
+
+      button = gtk_toggle_button_new ();
+      gtk_widget_set_tooltip_text (GTK_WIDGET (button), _("Show other locations"));
+      gtk_widget_set_focus_on_click (button, FALSE);
+      gtk_widget_set_valign (button, GTK_ALIGN_CENTER);
+      image = gtk_image_new_from_icon_name ("folder-symbolic", GTK_ICON_SIZE_MENU);
+      gtk_container_add (GTK_CONTAINER (button), image);
+      gtk_style_context_add_class (gtk_widget_get_style_context (button), "image-button");
+      gtk_style_context_remove_class (gtk_widget_get_style_context (button), "text-button");
+      gtk_widget_show (image);
+      gtk_widget_show (button);
+
+      header = gtk_dialog_get_header_bar (GTK_DIALOG (dialog));
+      gtk_header_bar_pack_start (GTK_HEADER_BAR (header), button);
+
+      sidebar = _gtk_file_chooser_widget_get_places (GTK_FILE_CHOOSER_WIDGET (dialog->priv->widget));
+      screen = gtk_widget_get_screen (GTK_WIDGET (dialog));
+      sidebar_visible = TRUE;
+
+      if (screen)
+        {
+          /* If the sidebar width takes more than 25% of the screen width, hide it */
+          gtk_widget_get_preferred_size (GTK_WIDGET (sidebar), &sidebar_requisition, NULL);
+          sidebar_visible = gdk_screen_get_width (screen) > sidebar_requisition.width * 4;
```

Comment:
This check should be the first thing you do in setup_places() and if the sidebar size is apropiatte for the screen, then we should not add the button to the headerbar